### DISCUSSION TITLE
Retrieving code completion/definitions from an external source

### DISCUSSION THREAD

**Discussion by marindevs** [python-lsp/python-lsp-server#445] [Q&A]
When I’m editing, I want to have access to some “built in” functions that should be available in the Editor (Completion and Definition) without importing.

The code for those built ins will be something like this:

```py
import numpy as np

myHeaderVar = 123

def myFunction(x):
    return np.abs(x)
```

And in the editor itself I will have:
```py

mylist = [1, 2, 3, 4]
myFunction(mylist)

some_value += myHeaderVar
```

This is just a silly example to capture the idea. When I start typing the myF.. it should provide the completion for the myFunction as it is present on the "built in headers".

So my question really is, there is something in the pylsp that support this? Like, loading this header.py from the same workspace that the main.py is and making the content of tha file available?
Or a way to provide the content of this Header.py to be used together with the code from main.py acting as dictionary for the completion and linting checking?

**Comment by lieryan** [comment]
You could create your own project-specific pylsp plugin. Disable the built-in completion plugin, make it so that it uses your completion plugin and your completion plugin will call the default completion and add whatever extras you need.

The [cookiecutter-pylsp-plugin](https://github.com/python-lsp/cookiecutter-pylsp-plugin) might be useful to you if you want to create your own plugin.
